diff --git a/gstreamer/pipeline.cpp b/gstreamer/pipeline.cpp
index ab53c10..3291c8b 100644
--- a/gstreamer/pipeline.cpp
+++ b/gstreamer/pipeline.cpp
@@ -30,6 +30,7 @@
 #include <gst/app/gstappsrc.h>
 #include <QtCore/QCoreApplication>
 #include <QtCore/QMutexLocker>
+#include <QtCore/QFile>
 #include "debug.h"
 #define MAX_QUEUE_TIME 20 * GST_SECOND
 
@@ -842,6 +843,12 @@ void Pipeline::cb_setupSource(GstElement *playbin, GParamSpec *param, gpointer d
             QString userAgent = QCoreApplication::applicationName() + '/' + QCoreApplication::applicationVersion();
             userAgent += QString(" (Phonon/%0; Phonon-GStreamer/%1)").arg(PHONON_VERSION_STR).arg(PHONON_GST_VERSION);
             g_object_set(phononSrc, "user-agent", userAgent.toUtf8().constData(), NULL);
+        } else if (that->currentSource().type() == MediaSource::Disc && 
+                g_object_class_find_property(G_OBJECT_GET_CLASS (phononSrc), "device")) {
+            QByteArray mediaDevice = QFile::encodeName(that->currentSource().deviceName());
+            if (!mediaDevice.isEmpty()) {
+                g_object_set(phononSrc, "device", mediaDevice.constData(), NULL);
+            }
         }
     }
     gst_object_unref(that->m_pipeline);
